Use schema when parsing csv

vvijay-bolt opened the pull request:
## What changes are proposed in this pull request?

Use the input schema when parsing csv input during model serving. This ensures that the resulting dataframe has the correct types.

## How is this patch tested?

Train and serve the modified logistic regression example below.
Save the following in test_csv.csv:
x
1
-1

Then try:
curl -d '{"columns":["x"],"index":[0,1],"data":[[1],[-1]]}' -H 'Content-Type: application/json' localhost:5000/invocations
curl --data-binary '@test_csv.csv' -H 'Content-Type: text/csv' localhost:5000/invocations

The json request returns [1, 0].
On master, the csv request will produce: {"error_code": "BAD_REQUEST", "message": "Incompatible input types for column x. Can not safely convert int64 to int32."}
After applying the fix, the csv request will also product [1, 0].

Modified logistic regression example:

import numpy as np
import pandas as pd
from sklearn.linear_model import LogisticRegression
from mlflow.models.signature import infer_signature

import mlflow
import mlflow.sklearn

if __name__ == "__main__":
    X = pd.DataFrame({'x': [-2, -1, 0, 1, 2, 1]}, dtype='int32')
    y = np.array([0, 0, 1, 1, 1, 0])
    lr = LogisticRegression()
    lr.fit(X, y)
    score = lr.score(X, y)
    print("Score: %s" % score)

    class Model(mlflow.pyfunc.PythonModel):
        def __init__(self, m):
            self.m = m

        def predict(self, context, x):
            return self.m.predict(x)

    model = Model(lr)
    signature = infer_signature(X, model.predict(None, X))
    print('Signature')
    print(signature)
    mlflow.pyfunc.log_model('model', python_model=model, signature=signature)
    print("Model saved in run %s" % mlflow.active_run().info.run_uuid)

## Does this PR change the documentation?

No

## Release Notes

### Is this a user-facing change?

No
### What component(s), interfaces, languages, and integrations does this PR affect?
Components 
- [ ] `area/model-registry`: Model Registry service, APIs, and the fluent client calls for Model Registry
- [ ] `area/server-infra`: MLflow Tracking server backend
- [ ] `area/tracking`: Tracking Service, tracking client APIs, autologging

<!--
Insert an empty named anchor here to allow jumping to this section with a fragment URL
(e.g. https://github.com/mlflow/mlflow/pull/123#user-content-release-note-category).
Note that GitHub prefixes anchor names in markdown with "user-content-".
-->
<a name="release-note-category"></a>
### How should the PR be classified in the release notes? Choose one:

- [- ] `rn/bug-fix` - A user-facing bug fix worth mentioning in the release notes

harupy committed:
Refactor files in `tests/db` to allow running tests in repository root (#5513)

* rename postgres.schema

Signed-off-by: harupy <hkawamura0130@gmail.com>

* Refactor database tests

Signed-off-by: harupy <hkawamura0130@gmail.com>

* remove import sqlalchemy

Signed-off-by: harupy <hkawamura0130@gmail.com>

* fix docker-compose down

Signed-off-by: harupy <hkawamura0130@gmail.com>

* update

Signed-off-by: harupy <hkawamura0130@gmail.com>

* fix lint failures

Signed-off-by: harupy <hkawamura0130@gmail.com>

* blacken

Signed-off-by: harupy <hkawamura0130@gmail.com>

* more fixes

Signed-off-by: harupy <hkawamura0130@gmail.com>

* add tests/db/compose.sh

Signed-off-by: harupy <17039389+harupy@users.noreply.github.com>

* rename

Signed-off-by: harupy <17039389+harupy@users.noreply.github.com>

* use compose v2

Signed-off-by: harupy <17039389+harupy@users.noreply.github.com>

* remove grep

Signed-off-by: harupy <hkawamura0130@gmail.com>

* run command using exec

Signed-off-by: harupy <hkawamura0130@gmail.com>

* set --no-TTY

Signed-off-by: harupy <hkawamura0130@gmail.com>

* specify image and remove redundant build arg

Signed-off-by: harupy <hkawamura0130@gmail.com>

* fix workflow

Signed-off-by: harupy <hkawamura0130@gmail.com>

* improve README

Signed-off-by: harupy <hkawamura0130@gmail.com>
Signed-off-by: Varun Kumar Vijay <vvijay@bolt.com>

singankit committed:
Append user agent added by request_header_provider to default user agent (#5502)

* User agent added by request_header_plugin gets appended to default user agent

Signed-off-by: singankit <anksing@microsoft.com>

* Adding large marker following similar test with the marker

Signed-off-by: singankit <anksing@microsoft.com>

* Appending headers from various request_header_providers

Signed-off-by: singankit <anksing@microsoft.com>

* Review comments to use dict method

Signed-off-by: singankit <anksing@microsoft.com>

* Adding default request header provider

Signed-off-by: singankit <anksing@microsoft.com>

* Fixing typo in test

Signed-off-by: singankit <anksing@microsoft.com>

* Fixing rest_utils test

Signed-off-by: singankit <anksing@microsoft.com>
Signed-off-by: Varun Kumar Vijay <vvijay@bolt.com>

dbczumar committed:
Use split (#5522)

Signed-off-by: dbczumar <corey.zumar@databricks.com>
Signed-off-by: Varun Kumar Vijay <vvijay@bolt.com>

bramrodenburg committed:
Fix issue #2804 - Set run name in MLProject execution (#5187)

* Add mlflow run name to projects.run()

Signed-off-by: bramrodenburg <14278376+bramrodenburg@users.noreply.github.com>

* Renamed MLFLOW_RUN_NAME to RUN_NAME in CLI metavars

Signed-off-by: bramrodenburg <14278376+bramrodenburg@users.noreply.github.com>

* Update CLI docstring

* Address feedback

* Autoformat: https://github.com/mlflow/mlflow/actions/runs/2031948113

Signed-off-by: mlflow-automation <mlflow-automation@users.noreply.github.com>

Co-authored-by: Corey Zumar <39497902+dbczumar@users.noreply.github.com>
Co-authored-by: mlflow-automation <mlflow-automation@users.noreply.github.com>
Signed-off-by: Varun Kumar Vijay <vvijay@bolt.com>

WeichenXu123 committed:
Restoring virtual python environment for mlflow pyfunc spark_udf (#5487)

* init

Signed-off-by: Weichen Xu <weichen.xu@databricks.com>

* update

Signed-off-by: Weichen Xu <weichen.xu@databricks.com>

* update

Signed-off-by: Weichen Xu <weichen.xu@databricks.com>

* update

Signed-off-by: Weichen Xu <weichen.xu@databricks.com>

* update

Signed-off-by: Weichen Xu <weichen.xu@databricks.com>

* update

Signed-off-by: Weichen Xu <weichen.xu@databricks.com>

* update

Signed-off-by: Weichen Xu <weichen.xu@databricks.com>

* update

Signed-off-by: Weichen Xu <weichen.xu@databricks.com>

* update

Signed-off-by: Weichen Xu <weichen.xu@databricks.com>

* debug ci

Signed-off-by: Weichen Xu <weichen.xu@databricks.com>

* update

Signed-off-by: Weichen Xu <weichen.xu@databricks.com>

* update

Signed-off-by: Weichen Xu <weichen.xu@databricks.com>

* update

Signed-off-by: Weichen Xu <weichen.xu@databricks.com>

* update

Signed-off-by: Weichen Xu <weichen.xu@databricks.com>

* update

Signed-off-by: Weichen Xu <weichen.xu@databricks.com>

* update

Signed-off-by: Weichen Xu <weichen.xu@databricks.com>

* black

Signed-off-by: Weichen Xu <weichen.xu@databricks.com>

* update

Signed-off-by: Weichen Xu <weichen.xu@databricks.com>

* update

Signed-off-by: Weichen Xu <weichen.xu@databricks.com>

* update

Signed-off-by: Weichen Xu <weichen.xu@databricks.com>

* update

Signed-off-by: Weichen Xu <weichen.xu@databricks.com>

* update

Signed-off-by: Weichen Xu <weichen.xu@databricks.com>

* update

Signed-off-by: Weichen Xu <weichen.xu@databricks.com>

* update

Signed-off-by: Weichen Xu <weichen.xu@databricks.com>

* update

Signed-off-by: Weichen Xu <weichen.xu@databricks.com>

* update

Signed-off-by: Weichen Xu <weichen.xu@databricks.com>

* fix tests

Signed-off-by: Weichen Xu <weichen.xu@databricks.com>

* update

Signed-off-by: Weichen Xu <weichen.xu@databricks.com>

* update

Signed-off-by: Weichen Xu <weichen.xu@databricks.com>

* fix test

Signed-off-by: Weichen Xu <weichen.xu@databricks.com>

* update

Signed-off-by: Weichen Xu <weichen.xu@databricks.com>

* update

Signed-off-by: Weichen Xu <weichen.xu@databricks.com>

* update

Signed-off-by: Weichen Xu <weichen.xu@databricks.com>

* update

Signed-off-by: Weichen Xu <weichen.xu@databricks.com>

* update

Signed-off-by: Weichen Xu <weichen.xu@databricks.com>

* update

Signed-off-by: Weichen Xu <weichen.xu@databricks.com>

* update

Signed-off-by: Weichen Xu <weichen.xu@databricks.com>

* update

Signed-off-by: Weichen Xu <weichen.xu@databricks.com>

* update

Signed-off-by: Weichen Xu <weichen.xu@databricks.com>

* update

Signed-off-by: Weichen Xu <weichen.xu@databricks.com>

* update

Signed-off-by: Weichen Xu <weichen.xu@databricks.com>

* update

Signed-off-by: Weichen Xu <weichen.xu@databricks.com>

* update

Signed-off-by: Weichen Xu <weichen.xu@databricks.com>

* update

Signed-off-by: Weichen Xu <weichen.xu@databricks.com>

* update

Signed-off-by: Weichen Xu <weichen.xu@databricks.com>

* update

Signed-off-by: Weichen Xu <weichen.xu@databricks.com>

* update

Signed-off-by: Weichen Xu <weichen.xu@databricks.com>

* update

Signed-off-by: Weichen Xu <weichen.xu@databricks.com>

* update

Signed-off-by: Weichen Xu <weichen.xu@databricks.com>

* update

Signed-off-by: Weichen Xu <weichen.xu@databricks.com>
Signed-off-by: Varun Kumar Vijay <vvijay@bolt.com>

vvijay-bolt committed:
use schema when parsing csv

Signed-off-by: Varun Kumar Vijay <vvijay@bolt.com>

harupy committed:
Set timezone to UTC in windows (#5538)

* Diable Daylight Saving time adjustments in windows

Signed-off-by: harupy <17039389+harupy@users.noreply.github.com>

* use utc

Signed-off-by: harupy <17039389+harupy@users.noreply.github.com>
Signed-off-by: Varun Kumar Vijay <vvijay@bolt.com>

WeichenXu123 committed:
init (#5532)

Signed-off-by: Weichen Xu <weichen.xu@databricks.com>
Signed-off-by: Varun Kumar Vijay <vvijay@bolt.com>

serena-ruan committed:
feat: pyspark autologging enhancement (#5481)

* feat: general support for autologging models containing parameters of type Param and fix some comments

Signed-off-by: Xinyue Ruan <serena.rxy@gmail.com>

* fix format

Signed-off-by: Xinyue Ruan <serena.rxy@gmail.com>

* fix jinja2 import error

Signed-off-by: Xinyue Ruan <serena.rxy@gmail.com>

* update function doc

Signed-off-by: Xinyue Ruan <serena.rxy@gmail.com>

* add Evaluator type param into testcase

Signed-off-by: Xinyue Ruan <serena.rxy@gmail.com>
Signed-off-by: Varun Kumar Vijay <vvijay@bolt.com>

BenWilson2 commented:
Hi @vvijay-bolt could you add a test suite to validate this change to ensure that the parsing of a .csv with different submitted data type orientations functions correctly? 
It can be added here: https://github.com/mlflow/mlflow/blob/master/tests/pyfunc/test_scoring_server.py

harupy committed:
Fix invalid Anaconda URLs (#5547)

Signed-off-by: harupy <17039389+harupy@users.noreply.github.com>
Signed-off-by: Varun Kumar Vijay <vvijay@bolt.com>

harupy committed:
Use latest version of black fo fix lint check (#5548)

* Use latest version of black

Signed-off-by: harupy <17039389+harupy@users.noreply.github.com>

* run black

Signed-off-by: harupy <17039389+harupy@users.noreply.github.com>
Signed-off-by: Varun Kumar Vijay <vvijay@bolt.com>

vinijaiswal committed:
Removing the mention of tf1. (#5552)

Removing the mention of tf1 since this directory mentions the examples of using tf2. For users on TF1, use the migration guide: https://www.tensorflow.org/guide/migrate

Signed-off-by: Varun Kumar Vijay <vvijay@bolt.com>

harupy committed:
Run `test_sqlalchemy_store.py` against PostgreSQL, MySQL, and MSSQL (#5540)

* run test_sqlalchemy_store

Signed-off-by: harupy <17039389+harupy@users.noreply.github.com>

* use IS NULL

Signed-off-by: harupy <17039389+harupy@users.noreply.github.com>

* rename test classes

Signed-off-by: harupy <17039389+harupy@users.noreply.github.com>

* fix like operation

Signed-off-by: harupy <17039389+harupy@users.noreply.github.com>

* use method

Signed-off-by: harupy <17039389+harupy@users.noreply.github.com>

* use constants

Signed-off-by: harupy <17039389+harupy@users.noreply.github.com>

* use LIKE before LIKE BINARY

Signed-off-by: harupy <17039389+harupy@users.noreply.github.com>

* use Japanese_Bushu_Kakusu

Signed-off-by: harupy <17039389+harupy@users.noreply.github.com>

* remove _UTF8

Signed-off-by: harupy <17039389+harupy@users.noreply.github.com>

* reset experiment_id

Signed-off-by: harupy <17039389+harupy@users.noreply.github.com>
Signed-off-by: Varun Kumar Vijay <vvijay@bolt.com>

WeichenXu123 committed:
Make tensorflow model prediction support array type input (#5545)

* init

Signed-off-by: Weichen Xu <weichen.xu@databricks.com>

* update

Signed-off-by: Weichen Xu <weichen.xu@databricks.com>

* update

Signed-off-by: Weichen Xu <weichen.xu@databricks.com>

* update

Signed-off-by: Weichen Xu <weichen.xu@databricks.com>

* add test

Signed-off-by: Weichen Xu <weichen.xu@databricks.com>

* update

Signed-off-by: Weichen Xu <weichen.xu@databricks.com>
Signed-off-by: Varun Kumar Vijay <vvijay@bolt.com>

vvijay-bolt committed:
test for csv parsing

Signed-off-by: Varun Kumar Vijay <vvijay@bolt.com>

vvijay-bolt commented:
@BenWilson2 I added a test (modeled after a similar json test) to verify that the datatypes were set correctly after parsing. The pandas functions to read and write csv files were throwing errors on the datetime fields, so I left them out of the test.

WeichenXu123 committed:
Implement function to get MLflow pyfunc model dependencies (#5503)

* init

Signed-off-by: Weichen Xu <weichen.xu@databricks.com>

* init

Signed-off-by: Weichen Xu <weichen.xu@databricks.com>

* update

Signed-off-by: Weichen Xu <weichen.xu@databricks.com>

* update

Signed-off-by: Weichen Xu <weichen.xu@databricks.com>

* update

Signed-off-by: Weichen Xu <weichen.xu@databricks.com>

* update

Signed-off-by: Weichen Xu <weichen.xu@databricks.com>

* update

Signed-off-by: Weichen Xu <weichen.xu@databricks.com>

* update

Signed-off-by: Weichen Xu <weichen.xu@databricks.com>

* update

Signed-off-by: Weichen Xu <weichen.xu@databricks.com>

* update

Signed-off-by: Weichen Xu <weichen.xu@databricks.com>

* update doc

Signed-off-by: Weichen Xu <weichen.xu@databricks.com>

* update

Signed-off-by: Weichen Xu <weichen.xu@databricks.com>

* split test

Signed-off-by: Weichen Xu <weichen.xu@databricks.com>

* update

Signed-off-by: Weichen Xu <weichen.xu@databricks.com>

* update

Signed-off-by: Weichen Xu <weichen.xu@databricks.com>

* update

Signed-off-by: Weichen Xu <weichen.xu@databricks.com>

* refactor

Signed-off-by: Weichen Xu <weichen.xu@databricks.com>

* update

Signed-off-by: Weichen Xu <weichen.xu@databricks.com>

* update

Signed-off-by: Weichen Xu <weichen.xu@databricks.com>

* update

Signed-off-by: Weichen Xu <weichen.xu@databricks.com>

* update

Signed-off-by: Weichen Xu <weichen.xu@databricks.com>

* update

Signed-off-by: Weichen Xu <weichen.xu@databricks.com>

* update

Signed-off-by: Weichen Xu <weichen.xu@databricks.com>
Signed-off-by: Varun Kumar Vijay <vvijay@bolt.com>

harupy committed:
replace universal_newlines with text (#5557)

Signed-off-by: harupy <17039389+harupy@users.noreply.github.com>
Signed-off-by: Varun Kumar Vijay <vvijay@bolt.com>

Kr4is committed:
Fix #5551 - PyTorch LightningDeprecationWarning ModelSummary and get_gpu_memory_map (#5559)

* Fix #5551 - PyTorch LightningDeprecationWarning ModelSummary and get_gpu_memory_map

Signed-off-by: Bruno Cabado Lousa <sr.brunocabado@gmail.com>

* Fix version check

Signed-off-by: Bruno Cabado Lousa <sr.brunocabado@gmail.com>
Signed-off-by: Varun Kumar Vijay <vvijay@bolt.com>

harupy committed:
Pin `click` to fix spacy tests (#5560)

* pin click

Signed-off-by: harupy <17039389+harupy@users.noreply.github.com>

* add spacy as extra dependency

Signed-off-by: harupy <17039389+harupy@users.noreply.github.com>

* pin flask

Signed-off-by: harupy <17039389+harupy@users.noreply.github.com>
Signed-off-by: Varun Kumar Vijay <vvijay@bolt.com>

arjundc-db commented on mlflow/utils/proto_json_utils.py:
```diff
@@ -193,6 +193,20 @@ def default(self, o):  # pylint: disable=E0202
             return super().default(o)
 
 
+def _extract_pandas_input_types_from_schema(schema):
```
If the input is of type csv, it has to be a pandas input. Since a pandas schema will always be a ColSpec, I dont think this entire function is necessary.
Simply using `dtypes = dict(zip(schema.input_names(), schema.pandas_types()))` should suffice.

vvijay-bolt replied:
Made the change.

apurva-koti committed:
Add default experiment ID provider for repo notebooks (#5449)

Signed-off-by: Varun Kumar Vijay <vvijay@bolt.com>

dogeplusplus committed:
Add feature to set and delete description of a run. (#5534)

* Add feature to set and delete description of a run.

Signed-off-by: Albert Chung <albertswchung@gmail.com>

* Add edge case handling when run description is set twice, add tests for each scenario.

Signed-off-by: Albert Chung <albertswchung@gmail.com>

* Throw INVALID_PARAMETER_VALUE error code when setting description twice.

Signed-off-by: Albert Chung <albertswchung@gmail.com>
Signed-off-by: Varun Kumar Vijay <vvijay@bolt.com>

vvijay-bolt committed:
Removed utility function, use zip

Signed-off-by: Varun Kumar Vijay <vvijay@bolt.com>

arjundc-db commented on mlflow/utils/proto_json_utils.py:
```diff
@@ -210,17 +210,7 @@ def _dataframe_from_json(
-        else:
-            dtypes = dict(zip(schema.input_names(), schema.pandas_types()))
-
+        dtypes = _extract_pandas_input_types_from_schema(schema)
```
Can you also revert this change please.

vvijay-bolt replied:
Done.

vvijay-bolt committed:
revert change to json utils

Signed-off-by: Varun Kumar Vijay <vvijay@bolt.com>

vvijay-bolt committed:
revert change to json utils

Signed-off-by: Varun Kumar Vijay <vvijay@bolt.com>

vvijay-bolt committed:
Merge branch 'master' into vvijay-bolt/fix_csv_parse

vvijay-bolt commented:
autoformat

mlflow-automation committed:
Autoformat: https://github.com/mlflow/mlflow/actions/runs/2091247297

Signed-off-by: mlflow-automation <mlflow-automation@users.noreply.github.com>

vvijay-bolt committed:
Merge branch 'master' into vvijay-bolt/fix_csv_parse

arjundc-db reviewed (approved):
LGTM.
Thank you @vvijay-bolt !

tomasatdatabricks reviewed (approved):
lgtm!

arjundc-db commented on mlflow/pyfunc/scoring_server/__init__.py:
```diff
@@ -136,14 +136,20 @@ def parse_json_input(json_input, orient="split", schema: Schema = None):
 
     try:
-        return pd.read_csv(csv_input)
+        csv_input = StringIO(csv_input)
```
Let's move this line back to  https://github.com/mlflow/mlflow/pull/5531/files#diff-e607d3e6aefa036b1bfe9416fd1aa8b0934120f458cf263ddebab224f832e4a5L264 since the input to parse_csv can also be a path to a csv file. We need to fix our documentation for this.

arjundc-db commented on tests/pyfunc/test_scoring_server.py:
```diff
@@ -429,6 +444,14 @@ def test_split_oriented_json_to_df():
+    schema = Schema([ColSpec(c, c) for c in pandas_df_with_csv_types.columns])
+    df = _shuffle_pdf(pandas_df_with_csv_types)
+    csv_str = df.to_csv(index=False)
+    df = pyfunc_scoring_server.parse_csv_input(csv_str, schema=schema)
```
We need StringIO here:
```suggestion
    df = pyfunc_scoring_server.parse_csv_input(csv_str, schema=schema)
```

Please import it using 
```
try:
    from StringIO import StringIO
except ImportError:
    from io import StringIO
```
to maintain compatibility.

arjundc-db reviewed (commented):
Sorry, two more comments to fix test failures and we should we good to go.

vvijay-bolt committed:
Merge branch 'master' into vvijay-bolt/fix_csv_parse

vvijay-bolt committed:
move string io

Signed-off-by: Varun Kumar Vijay <vvijay@bolt.com>

BenWilson2 commented on test_csv.csv:
```diff
@@ -0,0 +1,3 @@
+x
```
Where is this used?

vvijay-bolt committed:
remove file

Signed-off-by: Varun Kumar Vijay <vvijay@bolt.com>

vvijay-bolt committed:
remove files

Signed-off-by: Varun Kumar Vijay <vvijay@bolt.com>

BenWilson2 commented on examples/sklearn_logistic_mod/train.py:
```diff
@@ -0,0 +1,29 @@
+import numpy as np
```
Any new examples need to be registered to execute as part of the Examples CI here: https://github.com/mlflow/mlflow/blob/101ad6e8eb383c769178df0df83d1d2a1cea6b4a/tests/examples/test_examples.py#L92-L94
. Do we need this additional test or can a signature example be added to examples/sklearn_logistic_regression/train.py? 
It might also be worthwhile to add a load from the logged (saved) location to run the score inference again to demonstrate the loading of the pyfunc flavor with the type-validated signature for a .csv file.

BenWilson2 reviewed (approved):
LGTM!
